File: test/pass_test.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:passkite/pkpass/barcode.dart';
import 'package:passkite/pkpass/image.dart';
import 'package:passkite/pkpass/pass.dart';
import 'package:passkite/pkpass/pass_data.dart';
import 'package:passkite/pkpass/pass_field_content.dart';
import 'package:passkite/pkpass/pass_fields.dart';
import 'package:passkite/pkpass/personalize.dart';

void main() {
  group('Pass Generation Tests', () {
    late PassData basePassData;
    late String testCertificatePath;
    late String testPrivateKeyPath;
    late String testCertificatePem;
    late String testPrivateKeyPem;

    setUp(() {
      basePassData = PassData(
        description: 'Test Pass',
        organizationName: 'Test Organization',
        passTypeIdentifier: 'pass.com.example.test',
        serialNumber: '1234567890',
        teamIdentifier: 'ABCDE12345',
      );

      // These paths should be updated with actual test certificate and key paths
      testCertificatePath = 'test/certificates/test.pem';
      testPrivateKeyPath = 'test/certificates/test_key.pem';

      // For testing, we'll use dummy certificate and key data
      testCertificatePem =
          '-----BEGIN CERTIFICATE-----\nTEST\n-----END CERTIFICATE-----';
      testPrivateKeyPem =
          '-----BEGIN PRIVATE KEY-----\nTEST\n-----END PRIVATE KEY-----';
    });

    test('Generate basic generic pass', () {
      final pass = Pass(
        data: basePassData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      // Write to file for manual verification
      final file = File('test_generic_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Generic pass written to ${file.path}');
    });

    test('Generate boarding pass with all fields', () {
      final boardingPassData = PassData(
        description: basePassData.description,
        organizationName: basePassData.organizationName,
        passTypeIdentifier: basePassData.passTypeIdentifier,
        serialNumber: basePassData.serialNumber,
        teamIdentifier: basePassData.teamIdentifier,
        boardingPass: BoardingPass(
          transitType: TransitType.air,
          primaryFields: [
            PassFieldContent(
              key: 'departure',
              label: 'Departure',
              value: 'San Francisco',
            ),
            PassFieldContent(
              key: 'arrival',
              label: 'Arrival',
              value: 'New York',
            ),
          ],
          secondaryFields: [
            PassFieldContent(
              key: 'flight',
              label: 'Flight',
              value: 'AA123',
            ),
          ],
          auxiliaryFields: [
            PassFieldContent(
              key: 'gate',
              label: 'Gate',
              value: 'A12',
            ),
          ],
          backFields: [
            PassFieldContent(
              key: 'terms',
              label: 'Terms',
              value: 'Please arrive 2 hours before departure',
            ),
          ],
        ),
      );

      final pass = Pass(
        data: boardingPassData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      final file = File('test_boarding_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Boarding pass written to ${file.path}');
    });

    test('Generate coupon pass with barcode', () {
      final couponPassData = PassData(
        description: basePassData.description,
        organizationName: basePassData.organizationName,
        passTypeIdentifier: basePassData.passTypeIdentifier,
        serialNumber: basePassData.serialNumber,
        teamIdentifier: basePassData.teamIdentifier,
        coupon: Coupon(
          primaryFields: [
            PassFieldContent(
              key: 'discount',
              label: 'Discount',
              value: '20% OFF',
            ),
          ],
        ),
        barcodes: [
          Barcode(
            message: 'COUPON123',
            format: PkPassBarcodeType.qr,
            messageEncoding: 'iso-8859-1',
            altText: '20% OFF Coupon',
          ),
        ],
      );

      final pass = Pass(
        data: couponPassData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      final file = File('test_coupon_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Coupon pass written to ${file.path}');
    });

    test('Generate event ticket with images', () {
      // Create test images
      final logoImage = Uint8List.fromList([1, 2, 3, 4]); // Mock image data
      final thumbnailImage =
          Uint8List.fromList([5, 6, 7, 8]); // Mock image data

      final eventTicketData = PassData(
        description: basePassData.description,
        organizationName: basePassData.organizationName,
        passTypeIdentifier: basePassData.passTypeIdentifier,
        serialNumber: basePassData.serialNumber,
        teamIdentifier: basePassData.teamIdentifier,
        eventTicket: EventTicket(
          primaryFields: [
            PassFieldContent(
              key: 'event',
              label: 'Event',
              value: 'Summer Music Festival',
            ),
          ],
          secondaryFields: [
            PassFieldContent(
              key: 'date',
              label: 'Date',
              value: '2024-07-15',
            ),
          ],
        ),
      );

      final pass = Pass(
        data: eventTicketData,
        logo: PKImage(
          name: 'logo',
          images: {1: logoImage},
        ),
        thumbnail: PKImage(
          name: 'thumbnail',
          images: {1: thumbnailImage},
        ),
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      final file = File('test_event_ticket.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Event ticket written to ${file.path}');
    });

    test('Generate store card with personalization', () {
      final storeCardData = PassData(
        description: basePassData.description,
        organizationName: basePassData.organizationName,
        passTypeIdentifier: basePassData.passTypeIdentifier,
        serialNumber: basePassData.serialNumber,
        teamIdentifier: basePassData.teamIdentifier,
        storeCard: StoreCard(
          primaryFields: [
            PassFieldContent(
              key: 'balance',
              label: 'Balance',
              value: '\$50.00',
            ),
          ],
        ),
        userInfo: {
          'personalize': Personalize(
            description: 'Join our loyalty program',
            requiredPersonalizationFields: ['name', 'email'],
            termsAndConditions: 'Terms and conditions apply',
          ).toJson(),
        },
      );

      final pass = Pass(
        data: storeCardData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      final file = File('test_store_card.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Store card written to ${file.path}');
    });

    test('Pass generation with empty organization name', () {
      final invalidPassData = PassData(
        description: 'Test Pass',
        organizationName: '', // Empty organization name
        passTypeIdentifier: 'pass.com.example.test',
        serialNumber: '1234567890',
        teamIdentifier: 'ABCDE12345',
      );

      final pass = Pass(
        data: invalidPassData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      // Write to file for manual verification
      final file = File('test_empty_org_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Pass with empty organization name written to ${file.path}');
    });

    test('Pass generation with empty barcode message', () {
      final invalidPassData = PassData(
        description: 'Test Pass',
        organizationName: 'Test Organization',
        passTypeIdentifier: 'pass.com.example.test',
        serialNumber: '1234567890',
        teamIdentifier: 'ABCDE12345',
        barcodes: [
          Barcode(
            message: '', // Empty message
            format: PkPassBarcodeType.qr,
            messageEncoding: 'iso-8859-1',
            altText: 'Empty Barcode',
          ),
        ],
      );

      final pass = Pass(
        data: invalidPassData,
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      // Write to file for manual verification
      final file = File('test_empty_barcode_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Pass with empty barcode message written to ${file.path}');
    });

    test('Pass generation with invalid image data', () {
      final invalidImage =
          Uint8List.fromList([0, 1, 2, 3]); // Not a valid PNG image

      final invalidPassData = PassData(
        description: 'Test Pass',
        organizationName: 'Test Organization',
        passTypeIdentifier: 'pass.com.example.test',
        serialNumber: '1234567890',
        teamIdentifier: 'ABCDE12345',
      );

      final pass = Pass(
        data: invalidPassData,
        logo: PKImage(
          name: 'logo',
          images: {1: invalidImage},
        ),
      );

      final passBytes = pass.generate(
        certificatePem: null,
        privateKeyPem: null,
      );
      expect(passBytes, isNotNull);
      expect(passBytes!.length, greaterThan(0));

      // Write to file for manual verification
      final file = File('test_invalid_image_pass.pkpass');
      file.writeAsBytesSync(passBytes);
      print('Pass with invalid image data written to ${file.path}');
    });
  });
}

File: src/__tests__/Pass.test.ts
import { Pass, createPass } from '../Pass';
import { PassData, PassImage, PassImageType, BarcodeFormat, TransitType } from '../types';
import JSZip from 'jszip';

describe('Pass', () => {
  const minimalPassData: PassData = {
    formatVersion: 1,
    passTypeIdentifier: 'pass.com.example.test',
    serialNumber: 'TEST-001',
    teamIdentifier: 'ABCD1234',
    organizationName: 'Test Organization',
    description: 'Test Pass',
  };

  describe('createPass', () => {
    it('should create a new Pass instance', () => {
      const pass = createPass(minimalPassData);
      expect(pass).toBeInstanceOf(Pass);
    });

    it('should create Pass with images', () => {
      const images: PassImage[] = [
        { type: PassImageType.Icon, data: new Uint8Array([1, 2, 3]) },
      ];
      const pass = createPass(minimalPassData, images);
      expect(pass).toBeInstanceOf(Pass);
    });
  });

  describe('Pass constructor', () => {
    it('should accept pass data only', () => {
      const pass = new Pass(minimalPassData);
      expect(pass).toBeDefined();
    });

    it('should accept pass data with images', () => {
      const images: PassImage[] = [];
      const pass = new Pass(minimalPassData, images);
      expect(pass).toBeDefined();
    });

    it('should accept pass data with images and personalization', () => {
      const images: PassImage[] = [];
      const personalization = {
        requiredPersonalizationFields: [{ type: 'name' as const, description: 'Your name' }],
        description: 'Personalize your pass',
      };
      const pass = new Pass(minimalPassData, images, personalization);
      expect(pass).toBeDefined();
    });
  });

  describe('generate', () => {
    it('should generate a valid zip file without signature', async () => {
      const pass = createPass(minimalPassData);
      const buffer = await pass.generate({ skipSignature: true });

      expect(buffer).toBeInstanceOf(Uint8Array);
      expect(buffer.length).toBeGreaterThan(0);
    });

    it('should include pass.json in the zip', async () => {
      const pass = createPass(minimalPassData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJson = zip.file('pass.json');
      expect(passJson).not.toBeNull();
    });

    it('should include manifest.json in the zip', async () => {
      const pass = createPass(minimalPassData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const manifest = zip.file('manifest.json');
      expect(manifest).not.toBeNull();
    });

    it('should not include signature when skipSignature is true', async () => {
      const pass = createPass(minimalPassData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const signature = zip.file('signature');
      expect(signature).toBeNull();
    });

    it('should serialize pass data correctly', async () => {
      const passData: PassData = {
        ...minimalPassData,
        backgroundColor: 'rgb(255, 0, 0)',
        foregroundColor: 'rgb(255, 255, 255)',
        storeCard: {
          headerFields: [{ key: 'points', label: 'POINTS', value: '1000' }],
          primaryFields: [{ key: 'name', value: 'John Doe' }],
        },
      };

      const pass = createPass(passData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJsonFile = zip.file('pass.json');
      const passJsonContent = await passJsonFile!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.backgroundColor).toBe('rgb(255, 0, 0)');
      expect(parsed.storeCard.headerFields[0].key).toBe('points');
    });

    it('should include images in the zip', async () => {
      const iconData = new Uint8Array([137, 80, 78, 71]); // PNG magic bytes
      const images: PassImage[] = [
        { type: PassImageType.Icon, data: iconData },
      ];

      const pass = createPass(minimalPassData, images);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const icon = zip.file('icon.png');
      expect(icon).not.toBeNull();
    });

    it('should handle retina images correctly', async () => {
      const images: PassImage[] = [
        { type: PassImageType.Icon, scale: 1, data: new Uint8Array([1]) },
        { type: PassImageType.Icon, scale: 2, data: new Uint8Array([2]) },
        { type: PassImageType.Icon, scale: 3, data: new Uint8Array([3]) },
      ];

      const pass = createPass(minimalPassData, images);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      expect(zip.file('icon.png')).not.toBeNull();
      expect(zip.file('icon@2x.png')).not.toBeNull();
      expect(zip.file('icon@3x.png')).not.toBeNull();
    });

    it('should include all image types', async () => {
      const images: PassImage[] = [
        { type: PassImageType.Icon, data: new Uint8Array([1]) },
        { type: PassImageType.Logo, data: new Uint8Array([2]) },
        { type: PassImageType.Strip, data: new Uint8Array([3]) },
        { type: PassImageType.Background, data: new Uint8Array([4]) },
        { type: PassImageType.Footer, data: new Uint8Array([5]) },
        { type: PassImageType.Thumbnail, data: new Uint8Array([6]) },
      ];

      const pass = createPass(minimalPassData, images);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      expect(zip.file('icon.png')).not.toBeNull();
      expect(zip.file('logo.png')).not.toBeNull();
      expect(zip.file('strip.png')).not.toBeNull();
      expect(zip.file('background.png')).not.toBeNull();
      expect(zip.file('footer.png')).not.toBeNull();
      expect(zip.file('thumbnail.png')).not.toBeNull();
    });

    it('should include personalization.json when personalization is provided', async () => {
      const personalization = {
        requiredPersonalizationFields: [{ type: 'name' as const, description: 'Your name' }],
        description: 'Personalize your pass',
      };

      const pass = createPass(minimalPassData, [], personalization);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const personalizationFile = zip.file('personalization.json');
      expect(personalizationFile).not.toBeNull();

      const content = await personalizationFile!.async('string');
      const parsed = JSON.parse(content);
      expect(parsed.requiredPersonalizationFields).toHaveLength(1);
      expect(parsed.requiredPersonalizationFields[0].type).toBe('name');
    });

    it('should generate valid SHA1 hashes in manifest', async () => {
      const pass = createPass(minimalPassData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const manifestFile = zip.file('manifest.json');
      const manifestContent = await manifestFile!.async('string');
      const manifest = JSON.parse(manifestContent);

      // Manifest should have pass.json entry
      expect(manifest['pass.json']).toBeDefined();
      // SHA1 hash is 40 hex characters
      expect(manifest['pass.json']).toMatch(/^[a-f0-9]{40}$/);
    });

    it('should handle barcodes correctly', async () => {
      const passData: PassData = {
        ...minimalPassData,
        barcodes: [
          {
            format: BarcodeFormat.QR,
            message: 'https://example.com/pass/123',
            messageEncoding: 'iso-8859-1',
            altText: 'Scan me',
          },
        ],
      };

      const pass = createPass(passData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJsonFile = zip.file('pass.json');
      const passJsonContent = await passJsonFile!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.barcodes).toHaveLength(1);
      expect(parsed.barcodes[0].format).toBe(BarcodeFormat.QR);
      expect(parsed.barcodes[0].message).toBe('https://example.com/pass/123');
    });

    it('should handle Date objects in relevantDate', async () => {
      const passData: PassData = {
        ...minimalPassData,
        relevantDate: '2024-12-25T10:00:00.000Z',
      };

      const pass = createPass(passData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJsonFile = zip.file('pass.json');
      const passJsonContent = await passJsonFile!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.relevantDate).toBe('2024-12-25T10:00:00.000Z');
    });
  });

  describe('generateBase64', () => {
    it('should generate a valid base64 string', async () => {
      const pass = createPass(minimalPassData);
      const base64 = await pass.generateBase64({ skipSignature: true });

      expect(typeof base64).toBe('string');
      expect(base64.length).toBeGreaterThan(0);
    });

    it('should be decodable back to valid zip', async () => {
      const pass = createPass(minimalPassData);
      const base64 = await pass.generateBase64({ skipSignature: true });

      // Decode base64 to binary
      const binaryString = atob(base64);
      const bytes = new Uint8Array(binaryString.length);
      for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
      }

      // Verify it's a valid zip
      const zip = await JSZip.loadAsync(bytes);
      expect(zip.file('pass.json')).not.toBeNull();
    });
  });

  describe('setSigningCredentials', () => {
    it('should return the Pass instance for chaining', () => {
      const pass = createPass(minimalPassData);
      const result = pass.setSigningCredentials({
        wwdrCertificate: 'cert',
        signerCertificate: 'signer',
        signerKey: 'key',
      });
      expect(result).toBe(pass);
    });
  });

  describe('complex pass generation', () => {
    it('should generate a complete store card pass', async () => {
      const passData: PassData = {
        formatVersion: 1,
        passTypeIdentifier: 'pass.com.example.storecard',
        serialNumber: 'SC-001',
        teamIdentifier: 'TEAM123',
        organizationName: 'Example Store',
        description: 'Loyalty Card',
        logoText: 'EXAMPLE',
        backgroundColor: 'rgb(60, 65, 76)',
        foregroundColor: 'rgb(255, 255, 255)',
        labelColor: 'rgb(198, 202, 211)',
        storeCard: {
          headerFields: [
            { key: 'points', label: 'POINTS', value: '1,250' },
          ],
          primaryFields: [
            { key: 'name', label: 'MEMBER', value: 'John Appleseed' },
          ],
          secondaryFields: [
            { key: 'memberSince', label: 'MEMBER SINCE', value: '2024' },
            { key: 'level', label: 'LEVEL', value: 'Gold' },
          ],
          auxiliaryFields: [
            { key: 'balance', label: 'BALANCE', value: '$50.00' },
          ],
          backFields: [
            { key: 'terms', label: 'Terms', value: 'Terms and conditions apply.' },
          ],
        },
        barcodes: [
          {
            format: BarcodeFormat.QR,
            message: 'https://example.com/member/12345',
            messageEncoding: 'iso-8859-1',
          },
        ],
        locations: [
          { latitude: 37.7749, longitude: -122.4194, relevantText: 'Store Location' },
        ],
      };

      const images: PassImage[] = [
        { type: PassImageType.Icon, data: new Uint8Array([1, 2, 3]) },
        { type: PassImageType.Logo, data: new Uint8Array([4, 5, 6]) },
      ];

      const pass = createPass(passData, images);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);

      // Verify all expected files
      expect(zip.file('pass.json')).not.toBeNull();
      expect(zip.file('manifest.json')).not.toBeNull();
      expect(zip.file('icon.png')).not.toBeNull();
      expect(zip.file('logo.png')).not.toBeNull();

      // Verify pass.json content
      const passJsonContent = await zip.file('pass.json')!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.formatVersion).toBe(1);
      expect(parsed.storeCard).toBeDefined();
      expect(parsed.storeCard.headerFields).toHaveLength(1);
      expect(parsed.storeCard.primaryFields).toHaveLength(1);
      expect(parsed.storeCard.secondaryFields).toHaveLength(2);
      expect(parsed.barcodes).toHaveLength(1);
      expect(parsed.locations).toHaveLength(1);
    });

    it('should generate a boarding pass', async () => {
      const passData: PassData = {
        formatVersion: 1,
        passTypeIdentifier: 'pass.com.example.boarding',
        serialNumber: 'BP-001',
        teamIdentifier: 'TEAM123',
        organizationName: 'Example Airlines',
        description: 'Boarding Pass',
        boardingPass: {
          transitType: TransitType.Air,
          headerFields: [
            { key: 'gate', label: 'GATE', value: 'A12' },
          ],
          primaryFields: [
            { key: 'origin', label: 'SFO', value: 'San Francisco' },
            { key: 'destination', label: 'JFK', value: 'New York' },
          ],
          secondaryFields: [
            { key: 'passenger', label: 'PASSENGER', value: 'John Doe' },
          ],
          auxiliaryFields: [
            { key: 'flight', label: 'FLIGHT', value: 'EX123' },
            { key: 'seat', label: 'SEAT', value: '12A' },
          ],
        },
        barcodes: [
          {
            format: BarcodeFormat.PDF417,
            message: 'M1DOE/JOHN         EABC123 SFOJFKEX 1234 123Y012A0001 100',
          },
        ],
      };

      const pass = createPass(passData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJsonContent = await zip.file('pass.json')!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.boardingPass).toBeDefined();
      expect(parsed.boardingPass.transitType).toBe('PKTransitTypeAir');
      expect(parsed.barcodes[0].format).toBe(BarcodeFormat.PDF417);
    });

    it('should generate an event ticket', async () => {
      const passData: PassData = {
        formatVersion: 1,
        passTypeIdentifier: 'pass.com.example.event',
        serialNumber: 'ET-001',
        teamIdentifier: 'TEAM123',
        organizationName: 'Example Events',
        description: 'Concert Ticket',
        eventTicket: {
          primaryFields: [
            { key: 'event', label: 'EVENT', value: 'Summer Concert' },
          ],
          secondaryFields: [
            { key: 'date', label: 'DATE', value: 'July 15, 2024' },
            { key: 'venue', label: 'VENUE', value: 'City Arena' },
          ],
          auxiliaryFields: [
            { key: 'section', label: 'SECTION', value: 'A' },
            { key: 'row', label: 'ROW', value: '5' },
            { key: 'seat', label: 'SEAT', value: '12' },
          ],
        },
        semantics: {
          eventName: 'Summer Concert',
          venueName: 'City Arena',
        },
      };

      const pass = createPass(passData);
      const buffer = await pass.generate({ skipSignature: true });

      const zip = await JSZip.loadAsync(buffer);
      const passJsonContent = await zip.file('pass.json')!.async('string');
      const parsed = JSON.parse(passJsonContent);

      expect(parsed.eventTicket).toBeDefined();
      expect(parsed.semantics.eventName).toBe('Summer Concert');
    });
  });
});
